File: python/src/dexcost/session.py
"""Session-based auto-grouping for dexcost.

Groups related LLM and HTTP calls into a single task without
requiring explicit ``with dexcost.task():`` wrappers. Uses contextvars
for thread/async safety.
"""

from __future__ import annotations

import logging
import threading
import time
import uuid
from datetime import datetime, timezone

from dexcost.context import get_context, get_current_task, set_current_task
from dexcost.models.task import Task
from dexcost.storage.protocol import StorageBackend

_log = logging.getLogger(__name__)


class SessionManager:
    """Manages auto-created session tasks for grouping cost events.

    When no explicit ``dexcost.task()`` is active, the session manager
    creates a session task and sets it in the context so that subsequent
    LLM and HTTP calls are grouped together.
    """

    def __init__(self) -> None:
        self._sessions: dict[int, Task] = {}  # context id -> task
        self._last_activity: dict[int, float] = {}  # context id -> timestamp
        self._lock = threading.Lock()

    def get_or_create_session(
        self, call_type: str, storage: StorageBackend | None = None
    ) -> Task:
        """Return the active task or create a session task.

        If an explicit task is already active in the current context,
        that task is returned unchanged. Otherwise, a session task is
        created (or reused) for the current context.

        Args:
            call_type: Description of the call (e.g. ``"llm_call"``,
                ``"http_call"``).
            storage: Optional storage backend for persisting the task.

        Returns:
            The active or newly-created session task.
        """
        ctx_id = threading.get_ident()

        # If an explicit task is already active, use it. Automatic sessions
        # also live in the task ContextVar, so distinguish the managed session
        # for this thread from an explicit task before updating activity.
        existing = get_current_task()
        if existing is not None:
            with self._lock:
                managed = self._sessions.get(ctx_id)
                if existing is managed and existing.ended_at is None:
                    self._last_activity[ctx_id] = time.monotonic()
                    return existing
                if existing is managed:
                    self._sessions.pop(ctx_id, None)
                    self._last_activity.pop(ctx_id, None)
                elif existing.ended_at is None:
                    return existing

            # An automatic session can be finalized by the background sync
            # thread, which cannot clear another thread's ContextVar. Clear
            # the stale local reference before creating the next session.
            set_current_task(None)

        # Check if we already have a session for this thread/async context
        with self._lock:
            session = self._sessions.get(ctx_id)
            if session is not None and session.ended_at is None:
                self._last_activity[ctx_id] = time.monotonic()
                # Ensure it's set as current task
                set_current_task(session)
                return session
            if session is not None:
                self._sessions.pop(ctx_id, None)
                self._last_activity.pop(ctx_id, None)

        # Create a new session task
        ctx = get_context()
        task_id = uuid.uuid4()
        has_business_identity = ctx is not None and any(
            (
                ctx.customer_id,
                ctx.project_id,
                ctx.user_id,
                ctx.product_id,
                ctx.agent,
                ctx.workflow_id,
            )
        )
        session = Task(
            task_id=task_id,
            task_type="agent_session",
            status="pending",
            started_at=datetime.now(timezone.utc),
            customer_id=ctx.customer_id if ctx else None,
            project_id=ctx.project_id if ctx else None,
            user_id=ctx.user_id if ctx else None,
            product_id=ctx.product_id if ctx else None,
            root_task_id=task_id if has_business_identity else None,
            agent_id=ctx.agent if ctx else None,
            agent_version=ctx.agent_version if ctx else None,
            workflow_id=ctx.workflow_id if ctx else None,
            workflow_session_id=ctx.workflow_session_id if ctx else None,
            metadata=dict(ctx.metadata) if ctx and ctx.metadata else {},
        )

        if storage is not None:
            try:
                storage.insert_task(session)
            except Exception:
                _log.debug("Failed to persist session task", exc_info=True)

        with self._lock:
            self._sessions[ctx_id] = session
            self._last_activity[ctx_id] = time.monotonic()

        set_current_task(session)
        return session

    def finalize_idle_sessions(
        self,
        idle_seconds: float = 30.0,
        storage: StorageBackend | None = None,
    ) -> list[Task]:
        """Finalize sessions that have had no activity for *idle_seconds*.

        When *storage* is supplied, the finalized task is persisted before it
        is removed from the manager. ``update_task`` also marks a previously
        synced running snapshot pending again, allowing the final immutable
        business identity to be delivered by the next sync batch.

        Returns:
            List of finalized session tasks.
        """
        now = time.monotonic()
        finalized: list[Task] = []

        with self._lock:
            idle_ids = [
                ctx_id
                for ctx_id, last in self._last_activity.items()
                if (now - last) >= idle_seconds
            ]
            for ctx_id in idle_ids:
                session = self._sessions.get(ctx_id)
                if session is None:
                    self._last_activity.pop(ctx_id, None)
                    continue

                previous_status = session.status
                previous_ended_at = session.ended_at
                session.status = "success"
                session.ended_at = datetime.now(timezone.utc)

                if storage is not None:
                    try:
                        storage.update_task(session)
                    except Exception:
                        # Keep the session managed and retry persistence on a
                        # later sync cycle instead of losing its final state.
                        session.status = previous_status
                        session.ended_at = previous_ended_at
                        _log.warning(
                            "Failed to persist finalized session task",
                            exc_info=True,
                        )
                        continue

                self._sessions.pop(ctx_id, None)
                self._last_activity.pop(ctx_id, None)
                finalized.append(session)

        return finalized

    def clear(self) -> None:
        """Remove all tracked sessions (for testing)."""
        with self._lock:
            self._sessions.clear()
            self._last_activity.clear()


# Module-level singleton
_session_manager: SessionManager | None = None


def get_session_manager() -> SessionManager:
    """Return the global session manager, creating it if needed."""
    global _session_manager
    if _session_manager is None:
        _session_manager = SessionManager()
    return _session_manager


def reset_session_manager() -> None:
    """Reset the global session manager (for testing)."""
    global _session_manager
    if _session_manager is not None:
        _session_manager.clear()
    _session_manager = None

File: go/session.go
package dexcost

import (
	"context"
	"log"
	"sync"
	"time"

	"github.com/DexwoxBusiness/dexcost-sdk/go/core"
)

// SessionManager auto-groups HTTP/LLM calls that happen outside an
// explicit task context into session tasks. It is goroutine-safe.
//
// NOTE: Single mutex is acceptable for typical agent workloads (< 100 concurrent tasks).
// For extreme throughput (> 1000 concurrent), consider sharded locking or sync.Map.
type SessionManager struct {
	mu          sync.Mutex
	sessions    map[uint64]*sessionEntry
	idleTimeout time.Duration
	nextID      uint64
}

// sessionEntry tracks a session task and the last time it was used.
type sessionEntry struct {
	task         *core.Task
	lastActivity time.Time
	// identity groups consecutive identity-keyed calls (customer|project|agent)
	// into one session task. Empty for context-ID-keyed sessions.
	identity string
}

// NewSessionManager creates a SessionManager with the given idle timeout.
// Sessions that have not been active for idleTimeout will be finalized by
// FinalizeIdleSessions.
func NewSessionManager(idleTimeout time.Duration) *SessionManager {
	return &SessionManager{
		sessions:    make(map[uint64]*sessionEntry),
		idleTimeout: idleTimeout,
	}
}

// sessionContextKey is the context key for storing a session ID.
type sessionContextKey struct{}

// GetOrCreateSession returns the active task from ctx, or creates a new
// session task if none exists. The session task is stored in the manager
// and a derived context with the session task is returned.
//
// callType describes the initiating call (e.g. "llm_call", "http_call")
// and is stored in the session task's metadata. buffer is used to persist
// the session task; it may be nil if persistence is not desired.
func (sm *SessionManager) GetOrCreateSession(ctx context.Context, callType string, buffer core.Buffer) (context.Context, *core.Task) {
	// If an explicit task is already active, return it as-is.
	if existing := core.GetCurrentTask(ctx); existing != nil {
		sm.mu.Lock()
		// Update activity for all sessions (keyed by context session ID if present).
		if sid, ok := ctx.Value(sessionContextKey{}).(uint64); ok {
			if entry, found := sm.sessions[sid]; found {
				entry.lastActivity = time.Now()
			}
		}
		sm.mu.Unlock()
		return ctx, existing
	}

	// Check if context already carries a session ID.
	if sid, ok := ctx.Value(sessionContextKey{}).(uint64); ok {
		sm.mu.Lock()
		entry, found := sm.sessions[sid]
		if found {
			entry.lastActivity = time.Now()
			task := entry.task
			sm.mu.Unlock()
			return core.WithTask(ctx, task), task
		}
		sm.mu.Unlock()
	}

	// Create a new session task.
	taskType := "agent_session"
	if cd := core.GetContextData(ctx); cd != nil && cd.Agent != "" {
		taskType = cd.Agent
	}
	task := core.CreateAutoTask(ctx, taskType)
	task.Metadata["session"] = true
	task.Metadata["initiated_by"] = callType

	if buffer != nil {
		if err := buffer.InsertTask(task); err != nil {
			log.Printf("[dexcost] failed to persist session task: %v", err)
		}
	}

	sm.mu.Lock()
	sm.nextID++
	sid := sm.nextID
	sm.sessions[sid] = &sessionEntry{
		task:         &task,
		lastActivity: time.Now(),
	}
	sm.mu.Unlock()

	ctx = context.WithValue(ctx, sessionContextKey{}, sid)
	ctx = core.WithTask(ctx, &task)
	return ctx, &task
}

// GetOrCreateSessionForIdentity returns a session task that groups consecutive
// anonymous calls sharing the same (customer_id, project_id, agent) identity
// derived from the ambient DexcostContext. The HTTP adapter uses this so that
// a burst of HTTP calls rolls up into one session task instead of creating a
// throwaway task per request (Python parity: adapters/http.py groups by thread;
// Go has no goroutine identity, so we group by attribution identity instead).
// Sessions are reused until FinalizeIdleSessions closes them.
func (sm *SessionManager) GetOrCreateSessionForIdentity(ctx context.Context, callType string, buffer core.Buffer) *core.Task {
	// An explicit task always wins.
	if existing := core.GetCurrentTask(ctx); existing != nil {
		return existing
	}

	var customerID, projectID, agent string
	if cd := core.GetContextData(ctx); cd != nil {
		customerID, projectID, agent = cd.CustomerID, cd.ProjectID, cd.Agent
	}
	identity := customerID + "\x00" + projectID + "\x00" + agent

	// Sprint 2 Theme D / §3.2.2 (B13): single locked find-or-create
	// critical section. Pre-fix the lookup released the mutex before
	// running CreateAutoTask + buffer.InsertTask, then re-acquired to
	// insert — concurrent callers with the same identity all missed
	// the lookup, all created tasks, all inserted under separate IDs
	// → duplicate sessions on customer dashboards.
	//
	// The buffer.InsertTask call still happens INSIDE the lock. That
	// is acceptable: SQLite INSERTs are sub-ms locally and only ONE
	// task is ever persisted per identity per process lifetime. For
	// the unbounded-throughput case the file-level comment at line 15
	// recommends sharded locking.
	taskType := "agent_session"
	if agent != "" {
		taskType = agent
	}

	sm.mu.Lock()
	defer sm.mu.Unlock()

	for _, entry := range sm.sessions {
		if entry.identity == identity {
			entry.lastActivity = time.Now()
			return entry.task
		}
	}

	task := core.CreateAutoTask(ctx, taskType)
	task.Metadata["session"] = true
	task.Metadata["initiated_by"] = callType

	if buffer != nil {
		if err := buffer.InsertTask(task); err != nil {
			log.Printf("[dexcost] failed to persist session task: %v", err)
		}
	}

	// Sprint 4 §5.2 (A3) — hard count cap. The idleTimeout reaper
	// already evicts time-stale sessions, but a flood of distinct
	// identities (e.g. one anonymous user per request in a misconfigured
	// app) could otherwise exhaust process memory faster than the
	// reaper runs. Cap at 10 000 active sessions: when over, evict the
	// least-recently-active entry to make room.
	const sessionsCap = 10_000
	if len(sm.sessions) >= sessionsCap {
		var oldestID uint64
		var oldestActivity time.Time
		first := true
		for id, e := range sm.sessions {
			if first || e.lastActivity.Before(oldestActivity) {
				oldestID = id
				oldestActivity = e.lastActivity
				first = false
			}
		}
		delete(sm.sessions, oldestID)
	}

	sm.nextID++
	sm.sessions[sm.nextID] = &sessionEntry{
		task:         &task,
		lastActivity: time.Now(),
		identity:     identity,
	}
	return &task
}

// FinalizeIdleSessions closes sessions that have been idle for at least
// the configured idleTimeout. Finalized tasks are set to "success" status
// with an ended_at timestamp. If buffer is non-nil, each finalized task
// is updated in storage.
func (sm *SessionManager) FinalizeIdleSessions(buffer core.Buffer) []core.Task {
	now := time.Now()
	var finalized []core.Task

	sm.mu.Lock()
	for sid, entry := range sm.sessions {
		if now.Sub(entry.lastActivity) >= sm.idleTimeout {
			entry.task.Status = core.TaskStatusSuccess
			ended := time.Now().UTC()
			entry.task.EndedAt = &ended
			finalized = append(finalized, *entry.task)
			delete(sm.sessions, sid)
		}
	}
	sm.mu.Unlock()

	if buffer != nil {
		for _, task := range finalized {
			if err := buffer.UpdateTask(task); err != nil {
				log.Printf("[dexcost] failed to update finalized session task: %v", err)
			}
		}
	}

	return finalized
}

// ActiveSessionCount returns the number of currently active sessions.
func (sm *SessionManager) ActiveSessionCount() int {
	sm.mu.Lock()
	defer sm.mu.Unlock()
	return len(sm.sessions)
}

// Clear removes all tracked sessions. Intended for testing.
func (sm *SessionManager) Clear() {
	sm.mu.Lock()
	defer sm.mu.Unlock()
	sm.sessions = make(map[uint64]*sessionEntry)
}

// globalSessionManager is the package-level SessionManager singleton.
var (
	globalSessionManager     *SessionManager
	globalSessionManagerOnce sync.Once
)

// SessionMgr returns the global SessionManager, creating it if needed.
// The default idle timeout is 30 seconds.
func SessionMgr() *SessionManager {
	globalSessionManagerOnce.Do(func() {
		globalSessionManager = NewSessionManager(30 * time.Second)
	})
	return globalSessionManager
}

// resetSessionManager resets the global session manager. Used in tests.
func resetSessionManager() {
	if globalSessionManager != nil {
		globalSessionManager.Clear()
	}
	globalSessionManager = nil
	globalSessionManagerOnce = sync.Once{}
}
